contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,591 @@
+    }
+    return ctx;
+}
+
+
+
+static void waitUntilAllJobsCompleted(adaptCCtx* ctx)
+{
+    pthread_mutex_lock(&ctx->allJobsCompleted_mutex);
+    while (ctx->allJobsCompleted == 0) {
+        pthread_cond_wait(&ctx->allJobsCompleted_cond, &ctx->allJobsCompleted_mutex);
+    }
+    pthread_mutex_unlock(&ctx->allJobsCompleted_mutex);
+}
+
+static unsigned adaptCompressionLevel(adaptCCtx* ctx)
+{
+    unsigned reset = 0;
+    unsigned const allSlow = ctx->adaptParam < ctx->stats.compressedCounter && ctx->adaptParam < ctx->stats.writeCounter && ctx->adaptParam < ctx->stats.readyCounter ? 1 : 0;
+    unsigned const compressWaiting = ctx->adaptParam < ctx->stats.readyCounter ? 1 : 0;
+    unsigned const writeWaiting = ctx->adaptParam < ctx->stats.compressedCounter ? 1 : 0;
+    unsigned const createWaiting = ctx->adaptParam < ctx->stats.writeCounter ? 1 : 0;
+    unsigned const writeSlow = ((compressWaiting && createWaiting) || (createWaiting && !writeWaiting)) ? 1 : 0;
+    unsigned const compressSlow = ((writeWaiting && createWaiting) || (writeWaiting && !compressWaiting)) ? 1 : 0;
+    unsigned const createSlow = ((compressWaiting && writeWaiting) || (compressWaiting && !createWaiting)) ? 1 : 0;
+    DEBUGLOG(2, "ready: %u compressed: %u write: %u\n", ctx->stats.readyCounter, ctx->stats.compressedCounter, ctx->stats.writeCounter);
+    if (allSlow) {
+        reset = 1;
+    }
+    else if ((writeSlow || createSlow) && ctx->compressionLevel < (unsigned)ZSTD_maxCLevel()) {
+        DEBUGLOG(2, "increasing compression level %u\n", ctx->compressionLevel);
+        ctx->compressionLevel++;
+        reset = 1;
+    }
+    else if (compressSlow && ctx->compressionLevel > 1) {
+        DEBUGLOG(2, "decreasing compression level %u\n", ctx->compressionLevel);
+        ctx->compressionLevel--;
+        reset = 1;
+    }
+    if (reset) {
+        ctx->stats.readyCounter = 0;
+        ctx->stats.writeCounter = 0;
+        ctx->stats.compressedCounter = 0;
+    }
+    return ctx->compressionLevel;
+}
+
+static void* compressionThread(void* arg)
+{
+    adaptCCtx* ctx = (adaptCCtx*)arg;
+    unsigned currJob = 0;
+    for ( ; ; ) {
+        unsigned const currJobIndex = currJob % ctx->numJobs;
+        jobDescription* job = &ctx->jobs[currJobIndex];
+        DEBUGLOG(2, "compressionThread(): waiting on job ready\n");
+        pthread_mutex_lock(&ctx->jobReady_mutex);
+        while(currJob + 1 > ctx->jobReadyID) {
+            ctx->stats.waitReady++;
+            ctx->stats.readyCounter++;
+            DEBUGLOG(2, "waiting on job ready, nextJob: %u\n", currJob);
+            pthread_cond_wait(&ctx->jobReady_cond, &ctx->jobReady_mutex);
+        }
+        pthread_mutex_unlock(&ctx->jobReady_mutex);
+        DEBUGLOG(2, "compressionThread(): continuing after job ready\n");
+        /* compress the data */
+        {
+            unsigned const cLevel = adaptCompressionLevel(ctx);
+            DEBUGLOG(2, "cLevel used: %u\n", cLevel);
```

Comment:
minor detail : 
this `DEBUGLOG()` is incompatible with `-Wdeclaration-after-statement`, since a `size_t` variable is declared just after.
It works on Mac, because the default compiler basically ignore this compilation flag.
If you test it with gcc, or Linux, the compilation is likely to trigger a warning.

Note that respecting this compilation flag is not very important as long as your project remains in `/contrib`.